src/mame/sinclair/tsconf_copper.cpp

```diff
@@ -0,0 +1,217 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+/**********************************************************************
+    TS-Configuration Copper
+**********************************************************************/
+
+#include "emu.h"
+#include "tsconf_copper.h"
+
+
+#define LOG_CTRL  (1U << 1)
+#define LOG_DATA  (1U << 2)
+
+//#define VERBOSE ( LOG_GENERAL /*| LOG_CTRL | LOG_DATA*/ )
+#include "logmacro.h"
+
+#define LOGCTRL(...) LOGMASKED(LOG_CTRL, __VA_ARGS__)
+#define LOGDATA(...) LOGMASKED(LOG_DATA, __VA_ARGS__)
+
+
+// device type definition
+DEFINE_DEVICE_TYPE(TSCONF_COPPER, tsconf_copper_device, "tsconf_copper", "TS-Configuration Copper")
+
+
+tsconf_copper_device::tsconf_copper_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, TSCONF_COPPER, tag, owner, clock)
+	, m_timer(nullptr)
+	, m_cl_data(*this, "cl_data", 0x100, ENDIANNESS_LITTLE)
+	, m_out_wreg_cb(*this)
+	, m_in_dma_ready_cb(*this, 0)
+	, m_in_until_pos_cb(*this)
+{
+}
+
+void tsconf_copper_device::copper_en_w(u8 data)
+{
+	m_en = data != 0xff;
+	m_pc = data;
+
+	if (m_en)
+	{
+		LOGCTRL("START\n");
+		m_timer->adjust(attotime::zero);
+	}
+	else
+	{
+		LOGCTRL("STOP\n");
+		m_timer->reset();
+	}
+}
+
+void tsconf_copper_device::cl_data_w(u16 addr, u8 data)
+{
+	addr &= 0x1ff;
+	LOGDATA("data(W) %02x %04x\n", addr, data);
+	m_cl_data[addr >> 1] = (addr & 1)
+		? (m_cl_data[addr >> 1] & 0x00ff) | (data << 8)
+		: (m_cl_data[addr >> 1] & 0xff00) | data;
+}
+
+void tsconf_copper_device::dma_ready_w(int status)
+{
+	if(m_en && (status & 1) && (m_cl_data[m_pc] & 0xff0f) == 0b1111'0100'0000'0001)
+	{
+		++m_pc;
+		m_timer->adjust(attotime::from_hz(clock()), 0);
+	}
+}
+
+TIMER_CALLBACK_MEMBER(tsconf_copper_device::timer_callback)
+{
+	const u16 data = m_cl_data[m_pc];
+	if (BIT(data, 12, 4) == 0b1111) // all instructions except WREG
+	{
+		switch (BIT(data, 8, 4))
+		{
+			case 0b0110: //SETA
+				m_a = data & 0xff;
+				++m_pc;
+				break;
+
+			case 0b1000: //SETB
+			case 0b1001:
+				m_b = data & 0x1ff;
+				++m_pc;
+				break;
+
+			case 0b0100: // WAIT
+				{
+					const u8 event = data & 0x0f;
+					if (event == param)
+						++m_pc;
+					else
+					{
+						switch (event)
+						{
+							case 1: // DMA
+								m_timer->reset();
+								return;
+								break;
+							case 2: // frame
+								m_timer->adjust(m_in_until_pos_cb(0b00 << 14), 2);
+								return;
+								break;
+							case 3: // line start
+								m_timer->adjust(m_in_until_pos_cb(0b11 << 14), 3);
+								return;
+								break;
```

Comment:
Isn't return and break at same time going to throw a compiler error?

Reply:
Neither clang nor gcc failed the compilation. I was thinking having break everywhere is more consistent.